yabai: 3.3.10 -> 4.0.4

IvarWithoutBones opened the pull request:
###### Description of changes
This updates yabai to the latest version, which is required (from what I can tell) for it to function properly on modern MacOS versions. The old version used to flash and swap window positions constantly for me on 12.5.1, and that's entirely fixed with this update.

One thing I couldn't get to work natively is loading their scripting addition, which allows for more control over windows/spaces.
```shell
~/nix/nixpkgs > ./result/bin/yabai --load-sa                                                                                                                           
yabai: scripting-addition either failed to load or could not inject payload into Dock.app! Error: -1708
```
After looking into it i found those commands execute `/Library/ScriptingAdditions/yabai.osax/Contents/MacOS/mach_loader` under the hood, curiously doing that manually with `sudo mach_loader` loads the scripting addition just fine. That's not very intuitive though, so I wrote a script which does it for you which can be integrated into the home-manager/nix-darwin modules in the future. Notified upstream of the problem here: https://github.com/koekeishiya/yabai/discussions/1244#discussioncomment-3478453

I couldn't get cross compilation for aarch64-darwin to work, so ive marked it as broken on there. Highly doubt that's a regression anyways.

See [the changelog](https://github.com/koekeishiya/yabai/blob/v4.0.2/CHANGELOG.md) for more information.

cc @cmacrae @shardulbee

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and “core” functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

IvarWithoutBones committed:
yabai: 3.3.10 -> 4.0.2

IvarWithoutBones commented:
If you're testing this on aarch64-darwin you can apply the following patch:
```patch
diff --git a/pkgs/os-specific/darwin/yabai/default.nix b/pkgs/os-specific/darwin/yabai/default.nix
index f21eed3d516..9b0c56f4f5d 100644
--- a/pkgs/os-specific/darwin/yabai/default.nix
+++ b/pkgs/os-specific/darwin/yabai/default.nix
@@ -41,11 +41,13 @@ stdenv.mkDerivation rec {
   enableParallelBuilding = true;

   postPatch = ''
-    # aarch64 code is compiled on all targets, which causes Apple SDK headers to error out.
-    # Since multilib doesnt work on darwin i dont know of a better way of handling this.
     substituteInPlace makefile \
+    ${if stdenv.isx86_64 then ''
       --replace "-arch arm64e" "" \
       --replace "-arch arm64" ""
+    '' else if stdenv.isAarch64 then ''
+      --replace "-arch x86_64" ""
+    '' else throw "unsupported platform"}

     # `NSScreen::safeAreaInsets` is only available on macOS 12.0 and above, which frameworks arent packaged.
     # When a lower OS version is detected upstream just returns 0, so we can hardcode that at compiletime.
@@ -117,7 +119,7 @@ stdenv.mkDerivation rec {
     license = licenses.mit;
     platforms = platforms.darwin;
     # Fails to find `<Foundation/Foundation.h>` when cross compiling from x86_64-darwin, even with `CoreFoundation` added in `buildInputs`.
-    broken = hostPlatform.isAarch64;
+    #broken = hostPlatform.isAarch64;
     maintainers = with maintainers; [
       cmacrae
       shardy
```
Another user in an [upstream discussion thread](https://github.com/koekeishiya/yabai/discussions/1244#discussioncomment-3506974) reported compiling for `aarch64-darwin` doesn't work out of the box, based on [this discourse thread](https://discourse.nixos.org/t/error-unknown-target-cpu-armv8-3-a/16833) I think the patch might help.

epetousis commented:
Using this patch on aarch64 results in an `invalid arch name '-arch arm64e'` error when building the scripting extension, similar to what was reported in [this discourse thread](https://discourse.nixos.org/t/nixpkgs-aarch64-darwin-build-m1-uses-wrong-old-clang-how-to-fix/18368/2). It looks like the version of Clang being used (clang-11) is too old to support arm64e (which is weird to me, I thought the xcodebuild package wrapped around Xcode's clang?)

IvarWithoutBones commented:
> Using this patch on aarch64 results in an `invalid arch name '-arch arm64e'` error when building the scripting extension, similar to what was reported in [this discourse thread](https://discourse.nixos.org/t/nixpkgs-aarch64-darwin-build-m1-uses-wrong-old-clang-how-to-fix/18368/2). It looks like the version of Clang being used (clang-11) is too old to support arm64e

Ah, that's a great pointer!  I see that discourse thread just `substituteInPlaces` the architecture away, much like I do in this PR for `x86_64-darwin`. Maybe you could copy that line?

>  (which is weird to me, I thought the xcodebuild package wrapped around Xcode's clang?)


I believe it's because we're using `xcodebuild` from `xcbuild` instead of Xcode, because that isn't possible to provide purely. Looking at that projects homepage it appears its not been updated for a while, so it might ship with an older clang version? 

I'd try the following:
* Replace `stdenv` with `clang13Stdenv`, maybe `clang12Stdenv` if 13 doesn't work.
* Add this substitution so that we make sure to use our clang, and not the one from `xcbuild`:
```shell
substituteInPlace makefile --replace "xcrun clang" "clang"
```

IvarWithoutBones commented:
I was attempting to debug this the other day with @kira64xyz, where we got the following error after removing the `-arch arm64e` flag from the makefile:

```
yabai> ./src/osax/payload.m./src/osax/payload.m::3232::1010::  fatal errorfatal error: : ./src/osax/payload.m:32:10: 'ptrauth.h' file not found'ptrauth.h' file not found
yabai> fatal error
yabai> : 'ptrauth.h' file not found
yabai> #include <ptrauth.h>
yabai>          ^~~~~~~~~~~
yabai> #include <ptrauth.h>
yabai>          ^~~~~~~~~~~
yabai> #include <ptrauth.h>
yabai>          ^~~~~~~~~~~
yabai> 1 error generated.
yabai> 1 error generated.
yabai> 1 error generated.
yabai> make: *** [makefile:24: src/osax/loader_bin.c] Error 1
yabai> make: *** Waiting for unfinished jobs....
yabai> make: *** [makefile:24: src/osax/payload_bin.c] Error 1
yabai> make: *** [makefile:24: src/osax/mach_loader_bin.c] Error 1
error: builder for '/nix/store/kl9y3avqcbm6v0brc77rm9mcfk938vzs-yabai-4.0.2.drv' failed with exit code 2;
       last 10 log lines:
       >          ^~~~~~~~~~~
       > #include <ptrauth.h>
       >          ^~~~~~~~~~~
       > 1 error generated.
       > 1 error generated.
       > 1 error generated.
       > make: *** [makefile:24: src/osax/loader_bin.c] Error 1
       > make: *** Waiting for unfinished jobs....
       > make: *** [makefile:24: src/osax/payload_bin.c] Error 1
       > make: *** [makefile:24: src/osax/mach_loader_bin.c] Error 1
       For full logs, run 'nix log /nix/store/kl9y3avqcbm6v0brc77rm9mcfk938vzs-yabai-4.0.2.drv'.
```
After looking around Apples frameworks we found the `Kernel` package does indeed include the `ptrauth.h` header, but adding it to `buildInputs` didn't seem to help the situation as it would still fail to find the header.

We also tried adding the include using cflags as follows:
 ```nix
NIX_CFLAGS_COMPILE = [ "-I${Kernel}/Library/Frameworks/Kernel.framework/Headers/" ];
```
Sadly that only yielded less comprehensive and seemingly unrelated errors. I don't have the logs for those though, would you mind posting them @kira64xyz?


Figured I'd just post what we tried, maybe someone else can shed some light upon this. Or a more recent clang version just magically fixes it, one can hope :)

SuperSandro2000 commented on pkgs/os-specific/darwin/yabai/default.nix:
```diff
@@ -1,39 +1,127 @@
+
     cp ./bin/yabai $out/bin/yabai
+    ln -s ${loadScriptingAddition} $out/bin/yabai-load-sa
     cp ./doc/yabai.1 $out/share/man/man1/yabai.1
```
We should use installManPage instead.

SuperSandro2000 commented on pkgs/os-specific/darwin/yabai/default.nix:
```diff
@@ -1,39 +1,127 @@
+
+    loaderPath="/Library/ScriptingAdditions/yabai.osax/Contents/MacOS/mach_loader";
+
+    if ! test -f "$loaderPath" || ! test -x "$loaderPath"; then
```
```suggestion
    if ! test -x "$loaderPath"; then
```
Isn't that enough?

IvarWithoutBones replied:
It is 👍

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -36238,7 +36238,10 @@ with pkgs;
   xzoom = callPackage ../tools/X11/xzoom {};
 
-  yabai = callPackage ../os-specific/darwin/yabai { };
+  yabai = callPackage ../os-specific/darwin/yabai {
+    inherit (darwin.apple_sdk.frameworks) Cocoa Carbon ScriptingBridge;
+    inherit (darwin.apple_sdk_11_0.frameworks) SkyLight;
+  };
```
```suggestion
  yabai = darwin.apple_sdk_11_0.callPackage ../os-specific/darwin/yabai {
    inherit (darwin.apple_sdk.frameworks) Cocoa Carbon ScriptingBridge SkyLight;
  };
```

IvarWithoutBones replied:
What is the point of `apple_sdk_11.callPackage` if we still end up inheriting all the frameworks manually?

IvarWithoutBones replied:
Applied your suggestion, saw it uses some different libsystem stuff with the change. Assuming thats benefitial.

SuperSandro2000 replied:
It is just a thin wrapper around callPackage to make sure all the frameworks work and is required to use the new 11 sdk.

IvarWithoutBones commented:
> From what I understand arm64's pointer authentication support (and subsequently that ptrauth.h header) is actually baked into LLVM itself. Nix's xcodebuild wrapper is also using Nix's clang, which won't have this support baked in because [it hasn't been upstreamed yet.](https://reviews.llvm.org/D112941) I believe Apple's LLVM builds have it, though.

Hmm, that doesn't sound very trivial to fix. We could use a custom LLVM build with that patch but it doesn't really seem worth the effort and added complexity to me.

> The patch on the discourse thread straight up adds /usr/bin to PATH to use Xcode's clang, which is obviously not what we want.

Yeah, those impurities are definitely not something we want in nixpkgs. I switched to the binary release on aarch64-darwin for now, hopefully we can get this fixed in the future. PR should be ready to be merged now 👍

IvarWithoutBones commented:
I did also manage to get cross compilation from `x86_64-darwin` to `aarch64-darwin` to work for the from-source derivation, although I didn't really get much further than before. Left in the changes as it might help someone out in the future.

SuperSandro2000 commented on pkgs/os-specific/darwin/yabai/default.nix:
```diff
@@ -1,39 +1,179 @@
+    };
   };
-}
+}.${stdenv.hostPlatform.system}
```
```suggestion
}.${stdenv.hostPlatform.system} or (throw "unsupported system ${stdenv.hostPlatform.system}")
```

IvarWithoutBones replied:
Done.

sikmir commented on pkgs/os-specific/darwin/yabai/default.nix:
```diff
@@ -1,39 +1,179 @@
+let
   pname = "yabai";
-  version = "3.3.10";
+  version = "4.0.2";
```
4.0.4 is out

IvarWithoutBones replied:
Updated to it.

IvarWithoutBones committed:
yabai: 3.3.10 -> 4.0.4

IvarWithoutBones commented:
Anything blocking this? Would be great to get it merged.

IvarWithoutBones commented:
If someone on an `aarch64-darwin` system can confirm it works I'll merge this PR.

epetousis commented:
> If someone on an `aarch64-darwin` system can confirm it works I'll merge this PR.

I'm unable to disable SIP on my M1 Air currently so I can't test the scripting add-on, but this derivation builds without any issues and yabai works fine.

SuperSandro2000 commented:
@ofborg build yabai

Benjamin-L commented:
I tested this on my `aarch64-darwin` machine with SIP disabled, and the scripting addition doesn't seem to work. It throws the generic "failed to inject payload into Dock.app" error. When I attempt to run `mach_loader` by hand, it is immediately `SIGKILL`ed.

I'm _guessing_ that this is an upstream issue, since we're just downloading the prebuilt binaries for this architecture here.

IvarWithoutBones commented:
Would you mind testing if the script works on there? I'd guess it does not if launching `mach_loader` immediately SIGKILLs, since that's exactly what the script does. 

Just to make sure, are you launching it with root?

IvarWithoutBones commented:
I should also add, the "failed to inject payload into Dock.app" error message always shows up for me, while all the features needing the scripting additions work fine.

Benjamin-L commented:
The `yabai-load-sa` script has the same result, and yeah, in both cases I'm launching it as root.

After the installation attempt, I restarted yabai and then tested `yabai -m space --create`. It says "cannot create space due to an error with the scripting-addition", which I guess doesn't necessarily mean that the installation failed.

IvarWithoutBones commented:
The only occurance I had of this is when i updated yabai from 4.0.2 -> 4.0.4, where uninstalling SA and reinstalling it again fixed the issue. 

If that doesn't work I'm not quite sure what the problem is, we would need to dig through the source code. It might be attempting to write to the nix store or something? I cant think of many other things that would break this.

Benjamin-L commented:
Yeah, that's one of the things I tried, with the same result. Next step is probably to dtrace it to see if anything looks nix related. I can try that later today.

SuperSandro2000 commented:
No clue. If reinstalling it fixes it we should maybe change the wrapper to default to that but I don't know the implications of that.

mikezaby commented:
By the way there is a new major release 5.0.1.

ryanorendorff commented:
@IvarWithoutBones if you think this PR is ready to merge it would seem valuable to do so regardless of the 5.0.1 release; this PR most certainly moves yabai much closer to the more recent release. 

I tried to briefly compile yabai 5.0.1 with just a source change but there is a missing symbol error.

On Intel:

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/9nq5cb2gn1q1j0jbz3ypmcii371gq4al-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: -j12 SHELL=/nix/store/n63wlghd6cczckb9ymkgxz2pahlfvspk-bash-5.1-p16/bin/bash
rm -rf ./bin
xcrun clang ./src/osax/payload.m -shared -fPIC -O2 -mmacosx-version-min=11.0 -arch x86_64  -o ./src/osax/payload -F/System/Library/PrivateFrameworks -framework SkyLight -framework Foundation -framework Carbon
xcrun clang ./src/osax/payload.m -shared -fPIC -O2 -mmacosx-version-min=11.0 -arch x86_64  -o ./src/osax/payload -F/System/Library/PrivateFrameworks -framework SkyLight -framework Foundation -framework Carbon
Undefined symbols for architecture x86_64:
  "_objc_alloc_init", referenced from:
Undefined symbols for architecture x86_64:
  "_objc_alloc_init", referenced from:
      ___do_space_create_block_invoke in payload-706d50.o
      ___do_space_create_block_invoke in payload-3ecc5d.o
ld: symbol(s) not found for architecture x86_64
ld: symbol(s) not found for architecture x86_64
clang-11clang-11: : error: errorlinker command failed with exit code 1 (use -v to see invocation):
linker command failed with exit code 1 (use -v to see invocation)
make: *** [makefile:23: src/osax/loader_bin.c] Error 1
make: *** Waiting for unfinished jobs....
make: *** [makefile:23: src/osax/payload_bin.c] Error 1
```

SuperSandro2000 commented:
Regardless of the 5.0 update, is that PR mergeable in this state?

IvarWithoutBones commented:
> Regardless of the 5.0 update, is that PR mergeable in this state?

Yes. I have been using it for over a month now myself, and have noticed no instabilities. Seeing how the version of yabai on nixpkgs-unstable does not work at all on modern MacOS versions, I would definitely opt to merge this.

I currently do not have much time to work on this, so I will not be able to update this 5.0 for a while. I'll get on that as soon as I can, it can always be done in a followup PR.

IvarWithoutBones commented:
Completely forgot about my promise to merge this when someone confirmed it works on `aarch64-darwin`, I've done so now for the reasons stated in my previous comment.

If anyone gets around to updating this to the latest version, or fixing the scripting addition on `aarch64-darwin`, please don't hesitate to ping me in your PR.

IvarWithoutBones commented:
I've opened a PR which updates yabai to the latest version, figured id post it here in case people here would like to review or test it: #203504.
